Expand `ruff.configuration` to allow inline config (#16296)
astral-sh/ruff: docs/editors/settings.md

### `configuration`

> The `configuration` setting allows you to configure Ruff's behavior directly from an editor. This

Author: I don't think I follow with what you mean with "... to configure editor-specific Ruff." ?

> The `configuration` setting allows you to configure Ruff's behavior directly from an editor. This

Reviewer: oh, there's a missing *behavior* at the end.

can be done in one of the following ways:

1. **Configuration file path:** Specify the path to a `ruff.toml` or `pyproject.toml` file that
    contains the configuration. User home directory and environment variables will be expanded.
1. **Inline JSON configuration:** Directly provide the configuration as a JSON object. This is
    similar to how the configuration is specified in the settings tab on the
    [playground](https://play.ruff.sh/).

!!! note "Added in Ruff `0.9.8`"

    The **Inline JSON configuration** option was introduced in Ruff `0.9.8`, allowing you to define
    settings directly in your editor without relying on an external file.

By default, if no `configuration` is specified, Ruff will automatically detect and load settings

> from a configuration file in the project's filesystem, consistent with the behavior of the Ruff CLI.

Reviewer: I find this explanation a bit confusing for two reasons:

* You say **Ruff** will automatically detect.. consistent with the behavior of the *Ruff CLI*. What I find confusing about it is that it suggests that the editor integration is the "core" Ruff and the Ruff CLI is some form of an extension
* It's unclear to me what the project's filesystem is. 

The second item seems an issue inherent to the naming of our options. Naming the option `projectFirst` might have been more intuitive

Maybe;

The default behavior, if `configuration` is unset, is to load the settings from the project's configuration (a `ruff.toml` or `pyproject.toml` in the project's directory), consistent with when running Ruff on the CLI.

Author: I've incorporated this suggestion, thanks

> When both an editor-provided configuration and a local configuration file are present, you can
> control how Ruff resolves conflicts between them using the
> [`configurationPreference`](#configurationpreference) setting.

Reviewer: ```suggestion
When both an editor-provided configuration and a local configuration file are present, you can
control how Ruff resolves conflicts between them using the
The [`configurationPreference`](#configurationpreference) controls the precedence if both an editor-provided configuration (`ruff.configuration`) and a project level configuration are present.
```

#### Resolution order

In an editor, Ruff supports three sources of configuration, prioritized as follows (from highest to
lowest):

1. **Specific settings:** Individual settings like [`lineLength`](#linelength) or
    [`lint.select`](#select) defined in the editor
1. [**`ruff.configuration`**](#configuration): Settings provided via the
    [`configuration`](#configuration) field (either a path to a configuration file or an inline
    configuration object)
1. **Configuration file:** Settings defined in a `ruff.toml` or `pyproject.toml` file in the

>     project's filesystem (if present)

Reviewer: ```suggestion
    project's directory (if present)
```

> For example, if the line length is specified in all three sources, Ruff will use the value from the
> specific settings in the editor i.e., the [`lineLength`](#linelength) setting.

Reviewer: ```suggestion
For example, if the line length is specified in all three sources, Ruff will use the value from the [`lineLength`](#linelength) setting.
```

=== "VS Code"

    === "Using a configuration file"

        ```json
        {
            "ruff.configuration": "~/path/to/ruff.toml"
        }
        ```

    === "Using inline configuration"

        ```json
        {
            "ruff.configuration": {
                "lint": {
                    "unfixable": ["F401"],
                    "extend-select": ["TID251"],
                    "flake8-tidy-imports": {
                        "banned-api": {
                            "typing.TypedDict": {
                                "msg": "Use `typing_extensions.TypedDict` instead",
                            }
                        }
                    }
                },
                "format": {
                    "quote-style": "single"
                }
            }
        }

>         ```

Author: Yeah, it does seem simpler. Thanks for the suggestion!

<img width="2560" alt="Screenshot 2025-02-25 at 5 56 30 PM" src="https://github.com/user-attachments/assets/8a609655-7e7e-4edd-8366-3b9f562bf03c" />
